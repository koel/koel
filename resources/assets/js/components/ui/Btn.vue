<template>
  <button ref="button" type="button">
    <slot>Click me</slot>
  </button>
</template>

<script lang="ts" setup>
import { ref } from 'vue'

const button = ref<HTMLButtonElement>()

defineExpose({
  button
})
</script>

<style lang="scss" scoped>
button {
  background: var(--color-blue);
  color: var(--color-text-primary);
  font-size: 1rem;
  padding: .6rem .85rem;
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: .3rem;

  &:hover {
    box-shadow: inset 0 0 0 10rem rgba(0, 0, 0, .05);
  }

  &[small] {
    font-size: .9rem;
    padding: .4rem .7rem;
  }

  @include inset-when-pressed();

  &[green] {
    background-color: var(--color-green);
  }

  &[white] {
    background-color: transparent;
    color: var(--color-text-secondary);
  }

  &[red] {
    background-color: var(--color-red);
  }

  &[grey], &[gray] {
    background-color: var(--color-bg-secondary);
  }

  &[orange] {
    background-color: var(--color-highlight);
  }

  &[transparent] {
    background-color: transparent;
  }

  &[rounded] {
    border-radius: 999px;
  }

  &[unrounded] {
    border-radius: 0 !important;
  }

  &[uppercase] {
    text-transform: uppercase;
  }
}
</style>
