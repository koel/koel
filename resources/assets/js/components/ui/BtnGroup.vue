<template>
  <span class="btn-group">
    <slot></slot>
  </span>
</template>

<style lang="scss" scoped>
.btn-group {
  --radius: 5px;

  display: inline-flex;
  position: relative;
  flex-wrap: nowrap;

  :deep(button) {
    &:not(:first-child) {
      border-radius: 0;
    }

    &:first-of-type {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
      border-top-left-radius: var(--radius);
      border-bottom-left-radius: var(--radius);
    }

    &:last-of-type {
      border-top-right-radius: var(--radius);
      border-bottom-right-radius: var(--radius);
    }
  }

  &[uppercased] :deep(button) {
    text-transform: uppercase;
  }
}
</style>
