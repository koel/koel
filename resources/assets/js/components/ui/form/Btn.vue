<template>
  <button
    v-if="tag === 'button'"
    ref="button"
    class="text-base text-white border border-transparent bg-k-primary px-3.5 py-2 rounded cursor-pointer"
    type="button"
  >
    <slot>Click me</slot>
  </button>
  <a
    v-else
    ref="button"
    class="text-base text-white border border-transparent bg-k-primary px-3.5 py-2 rounded cursor-pointer"
  >
    <slot>Click me</slot>
  </a>
</template>

<script lang="ts" setup>
import { ref } from 'vue'

withDefaults(defineProps<{ tag?: 'button' | 'a' }>(), {
  tag: 'button',
})

const button = ref<HTMLButtonElement>()

defineExpose({
  button,
})
</script>

<style lang="postcss" scoped>
/**
 * Except for the `highlight` variant, button text colors are independent from the theme.
 */
button,
a {
  &:not([disabled]):hover {
    box-shadow: inset 0 0 0 10rem rgba(0, 0, 0, 0.1);
  }

  &:not([disabled]):active {
    box-shadow: inset 0 10px 10px -10px rgba(0, 0, 0, 0.6);
  }

  &[big] {
    @apply px-6 py-3;
  }

  &[small] {
    @apply text-[0.9rem] px-3 py-1;
  }

  &[success] {
    @apply bg-k-success text-white;
  }

  &[white],
  &[transparent],
  &[gray] {
    @apply bg-transparent text-k-fg hover:text-k-fg-80 active:text-k-fg-70;
  }

  &[danger] {
    @apply bg-k-danger text-white;
  }

  &[highlight] {
    @apply bg-k-highlight text-k-highlight-fg;
  }

  &[rounded] {
    @apply rounded-full;
  }

  &[unrounded] {
    @apply rounded-none;
  }

  &[uppercase] {
    @apply uppercase;
  }

  &[bordered] {
    @apply border-k-fg-20;
  }
}
</style>
