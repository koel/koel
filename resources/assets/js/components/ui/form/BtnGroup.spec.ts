import { expect, it } from 'vitest'
import UnitTestCase from '@/__tests__/UnitTestCase'
import BtnGroup from './BtnGroup.vue'
import Btn from './Btn.vue'

new class extends UnitTestCase {
  protected test () {
    it('renders', () => {
      expect(this.render(BtnGroup, {
        slots: {
          default: ['Green', 'Orange', 'Blue'].map(text => this.renderButtonToSlot(text)),
        },
      }).html()).toMatchSnapshot()
    })
  }

  private renderButtonToSlot (text: string) {
    return this.render(Btn, {
      slots: {
        default: text,
      },
    }).html()
  }
}
