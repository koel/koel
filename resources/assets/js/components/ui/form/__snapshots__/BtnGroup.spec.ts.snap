// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`btnGroup.vue > renders 1`] = `<span data-v-cf9b67d8="" class="btn-group inline-flex relative flex-nowrap"><button data-v-8943c846="" class="text-base text-white border border-transparent bg-k-primary px-3.5 py-2 rounded cursor-pointer" type="button">Green</button><button data-v-8943c846="" class="text-base text-white border border-transparent bg-k-primary px-3.5 py-2 rounded cursor-pointer" type="button">Orange</button><button data-v-8943c846="" class="text-base text-white border border-transparent bg-k-primary px-3.5 py-2 rounded cursor-pointer" type="button">Blue</button></span>`;
